### PR TITLE
Add unit tests for terms aggregation objects.

### PR DESCRIPTION
**Pull request by jpountz**
Relates #22278

### CODE REVIEW

**Comment by jpountz** [comment]
@elasticmachine retest this please

**Comment by jpountz** [comment]
@elasticmachine test this please

**Comment by jpountz** [comment]
@elasticmachine retest this please

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/DoubleTerms.java [hunk 1]
```diff
@@ -99,6 +100,16 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
             builder.endObject();
             return builder;
         }
+
+        @Override
+        public boolean equals(Object obj) {
+            return super.equals(obj) && Objects.equals(term, ((Bucket) obj).term);
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/DoubleTerms.java]: This probably needs the traditional class comparison to make sure the cast doesn't blow up.

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalMappedTerms.java [hunk 2]
```diff
@@ -110,4 +111,20 @@ public B getBucketByKey(String term) {
         }
         return bucketMap.get(term);
     }
+
+    @Override
+    protected boolean doEquals(Object obj) {
+        InternalMappedTerms<?,?> that = (InternalMappedTerms<?,?>) obj;
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalMappedTerms.java]: Here too.

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalTerms.java [hunk 3]
```diff
@@ -269,4 +289,17 @@ public InternalAggregation doReduce(List<InternalAggregation> aggregations, Redu
      * Create an array to hold some buckets. Used in collecting the results.
      */
     protected abstract B[] createBucketsArray(int size);
+
+    @Override
+    protected boolean doEquals(Object obj) {
+        InternalTerms<?,?> that = (InternalTerms<?,?>) obj;
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalTerms.java]: Here too.

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/LongTerms.java [hunk 4]
```diff
@@ -99,6 +100,16 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
             builder.endObject();
             return builder;
         }
+
+        @Override
+        public boolean equals(Object obj) {
+            return super.equals(obj) && Objects.equals(term, ((Bucket) obj).term);
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/LongTerms.java]: Again.

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/StringTerms.java [hunk 5]
```diff
@@ -95,6 +96,16 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
             builder.endObject();
             return builder;
         }
+
+        @Override
+        public boolean equals(Object obj) {
+            return super.equals(obj) && Objects.equals(termBytes, ((Bucket) obj).termBytes);
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/StringTerms.java]: Another.

File: core/src/test/java/org/elasticsearch/search/aggregations/bucket/terms/DoubleTermsTests.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        int shardSize = requiredSize + 2;
+        DocValueFormat format = DocValueFormat.RAW;
+        boolean showTermDocCountError = false;
+        long docCountError = -1;
+        long otherDocCount = 0;
+        List<DoubleTerms.Bucket> buckets = new ArrayList<>();
+        final int numBuckets = randomInt(shardSize);
+        Set<Double> terms = new HashSet<>();
+        for (int i = 0; i < numBuckets; ++i) {
+            double term;
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/aggregations/bucket/terms/DoubleTermsTests.java]: `randomValueOtherThanMany` might make this easier to read.

**Review by nik9000 — APPROVED** [review]

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/DoubleTerms.java (hunk 1 above)
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/DoubleTerms.java]: The super.equals call already guarantees it.

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalMappedTerms.java (hunk 2 above)
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalMappedTerms.java]: It is called from equals, which already does this check.

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/DoubleTerms.java (hunk 1 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/DoubleTerms.java]: Ah, fine then!

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalMappedTerms.java (hunk 2 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalMappedTerms.java]: Yeah, I see now. Sorry, I wasn't being careful.

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/StringTerms.java (hunk 5 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/StringTerms.java]: This one is fine too.

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/LongTerms.java (hunk 4 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/LongTerms.java]: As is this one.

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalTerms.java (hunk 3 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalTerms.java]: And this one is also fine.

**Pull request merged**